federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+	//we need two clientsets to talk to federation apiserver now, federationClient talks to cluster
+	//kubeclient talks to services.
+	federationClient   federationclientset.Interface
+	federationName     string
+	clusterClientMap   map[string]clientset.Clientset
+	balancer           cloudprovider.LoadBalancer
+	zones              []dnsprovider.Zone
+	cache              *serviceCache
+	clusterClientCache *clusterClientCache
+	// A store of services, populated by the serviceController
+	serviceStore       cache.StoreToServiceLister
+	// Watches changes to all services
+	serviceController  *framework.Controller
+	// A store of services, populated by the serviceController
+	clusterStore       federationcache.StoreToClusterLister
+	// Watches changes to all services
+	clusterController  *framework.Controller
+	eventBroadcaster   record.EventBroadcaster
+	eventRecorder      record.EventRecorder
+	//clusterLister     federationcache.StoreToClusterLister
+	// services that need to be synced
+	queue              *workqueue.Type
+}
+
+// New returns a new service controller to keep cloud provider service resources
+// (like Kubernetes Services and DNS server records for service discovery) in sync with the registry.
+
+func New(federationClient federationclientset.Interface, dns dnsprovider.Interface, federationName string) *ServiceController {
+	// TODO: quinton:
+	//                 1. cloudprovider.Interface is used for DNS configuration?
+	//                 2. need one kubeClient per cluster, or look up based on type of cluster.
+	//                 3. Replace clustername with cluster api object reference.
+	broadcaster := record.NewBroadcaster()
+	//broadcaster.StartRecordingToSink(&unversioned_core.EventSinkImpl{Interface: kubeClient.Core().Events("")})
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "federated-service-controller"})
+
+	sc := &ServiceController{
+		dns:              dns,
+		knownClusterSet:  make(sets.String),
+		federationClient: federationClient,
+		clusterClientMap: make(map[string]clientset.Clientset),
+		federationName:   federationName,
+		cache:            &serviceCache{serviceMap: make(map[string]*cachedService)},
+		clusterClientCache: &clusterClientCache{ clientMap: make(map[string]*cachedClusterClient)},
+		eventBroadcaster: broadcaster,
+		eventRecorder:    recorder,
+		queue: workqueue.New(),
+	}
+	sc.serviceStore.Store, sc.serviceController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return sc.federationClient.Core().Services(api.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return sc.federationClient.Core().Services(api.NamespaceAll).Watch(options)
+			},
+		},
+		&api.Service{},
+		serviceSyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: sc.enqueueService,
+			UpdateFunc: func(old, cur interface{}) {
+				sc.enqueueService(cur)
```

Comment:
I think we might want both the old and the current here, so that we can execute a diff/patch operation.